src/librustc_parse/parser/item.rs

```diff
@@ -640,6 +643,13 @@ impl<'a> Parser<'a> {
                 )
             }
             None => {
+                // Reject `impl const Type {}` here
+                if let Some(Spanned { node: Constness::Const, span }) = constness {
+                    self.struct_span_err(span, "`const` cannot modify an inherent impl")
+                        .help("only a trait impl can be `const`")
+                        .emit();
+                }
+
```

Comment:
If they remain, I'd like to move the checks for `?const` in trait objects and `impl Trait` into `ast_validation` as well. Currently they are done during lowering which is suboptimal.

Reply:
Sounds fine to me. :)

Reply:
Actually we shouldn't move this particular check into AST validation, since `const Trait` should never become a trait object. We should reject this in the parser. The other checks were moved to AST validation.